test/e2e/app_multiple_sources_test.go

```diff
@@ -167,3 +167,62 @@ func TestMultiSourceAppWithSourceOverride(t *testing.T) {
 			assert.Contains(t, output, "foo=bar")
 		})
 }
+
+func TestMultiSourceAppWithSourceName(t *testing.T) {
+	sources := []ApplicationSource{{
+		RepoURL: RepoURL(RepoURLTypeFile),
+		Path:    guestbookPath,
+		Name:    "guestbook",
+	}, {
+		RepoURL: RepoURL(RepoURLTypeFile),
+		Path:    "two-nice-pods",
+		Name:    "dynamic duo",
+	}}
+	ctx := Given(t)
+	ctx.
+		Sources(sources).
+		When().
+		CreateMultiSourceAppFromFile().
+		Then().
+		And(func(app *Application) {
+			assert.Equal(t, Name(), app.Name)
+			for i, source := range app.Spec.GetSources() {
+				assert.Equal(t, sources[i].RepoURL, source.RepoURL)
+				assert.Equal(t, sources[i].Path, source.Path)
+				assert.Equal(t, sources[i].Name, source.Name)
+			}
+			assert.Equal(t, DeploymentNamespace(), app.Spec.Destination.Namespace)
+			assert.Equal(t, KubernetesInternalAPIServerAddr, app.Spec.Destination.Server)
+		}).
+		Expect(Event(EventReasonResourceCreated, "create")).
+		And(func(app *Application) {
+			// we remove the first source
+			output, err := RunCli("app", "remove-source", Name(), "--source-name", "guestbook")
+			require.NoError(t, err)
+			assert.Contains(t, output, "updated successfully")
+			assert.Len(t, app.Spec.GetSources(), 2)
+		}).
+		Expect(Success("")).
+		And(func(app *Application) {
+			// we add a source
+			output, err := RunCli("app", "add-source", Name(), "--source-name", "guestbook2", "--repo", RepoURL(RepoURLTypeFile), "--path", guestbookPath)
+			require.NoError(t, err)
+			assert.Contains(t, output, "updated successfully")
+			assert.Len(t, app.Spec.GetSources(), 1)
+		}).
+		Expect(Success("")).
+		Given().Timeout(60).
+		When().Wait().Then().
+		Expect(Success("")).
+		And(func(app *Application) {
+			statusByName := map[string]SyncStatusCode{}
+			for _, r := range app.Status.Resources {
+				statusByName[r.Name] = r.Status
+			}
+			// check if the app has 3 resources, guestbook and 2 pods
```

Comment:
I think you should validate that the final source is equal to two and has "dynamic duo" and "guestbook2".